backend/uclapi/oauth/views.py

```diff
@@ -0,0 +1,347 @@
+from django.core.signing import TimestampSigner
+from django.utils.http import quote
+from django.core.serializers.json import DjangoJSONEncoder
+from django.views.decorators.csrf import csrf_exempt, ensure_csrf_cookie, csrf_protect
+import requests
+
+import os
+import json
+
+from dashboard.models import App, User
+from .app_helpers import generate_random_verification_code
+from .models import OAuthScope, OAuthToken
+
+# The endpoint that creates a Shibboleth login and redirects the user to it
+def authorise(request):
+    client_id = request.GET.get("client_id", None)
+    state = request.GET.get("state", None)
+    if not (client_id and state):
+        return JsonResponse({
+            "ok": False,
+            "error": "incorrect parameters supplied"
+        })
+    
+    try:
+        app = App.objects.get(client_id=client_id)
+    except ObjectDoesNotExist:
+        return JsonResponse({
+            "ok": False,
+            "error": "App does not exist for client id"
+        })
+
+    # Sign the app and state pair before heading to Shibboleth to help protect
+    # against CSRF and XSS attacks
+    signer = TimestampSigner()
+    data = app.client_id + "|" + state
+    signed_data = signer.sign(data)
+
+    # Build Shibboleth callback URL
+    url = os.environ.get("SHIBBOLETH_ROOT") + "/Login?target="
+    target = request.build_absolute_uri(request.path)
+    target += "shibcallback?appdata="
+    target += signed_data
+    target = quote(target)
+    url += target
+
+    # Send the user to Shibboleth to log in
+    return redirect(url)
+
+@csrf_exempt
+@ensure_csrf_cookie
+def shibcallback(request):
+    # Callback from Shib login. Get ALL the meta!
+    appdata_signed = request.GET.get("appdata", None)
+    if not appdata_signed:
+        return JsonResponse({
+            "ok": False,
+            "error": "No signed app data returned from Shibboleth. Please use the authorise endpoint."
+        })
+
+    signer = TimestampSigner()
+    try:
+        # Expire our signed tokens after five minutes for added security
+        appdata = signer.unsign(appdata_signed, max_age=300)
+    except signing.BadSignature:
+        return JsonResponse({
+            "ok": False,
+            "error": "Bad signature. Please try login again."
+        })
+    except signing.SignatureExpired:
+        return JsonResponse({
+            "ok": False,
+            "error": "Signature has expired. Please try login again."
+        })
+    
+    parts = appdata.split("|")
+    client_id = parts[0]
+    state = parts[1]
+
+    app = App.objects.get(client_id=client_id)
+
+    try:
+        eppn = request.META['HTTP_EPPN']
+        groups = request.META['HTTP_UCLINTRANETGROUPS']
+        cn = request.META['HTTP_CN']
+        department = request.META['HTTP_DEPARTMENT']
+        given_name = request.META['HTTP_GIVENNAME']
+        display_name = request.META['HTTP_DISPLAYNAME']
+        employee_id = request.META['HTTP_EMPLOYEEID']
+    except:
```

Comment:
Is there a list of exceptions to catch for each function? I'm crappy at Python as I'm sure you can now see!